src/tracker.rs

```diff
@@ -0,0 +1,199 @@
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    iter::FromIterator,
+    sync::{Mutex, MutexGuard},
+};
+
+use crate::Datadog;
+
+/// See https://www.datadoghq.com/pricing/ and https://docs.datadoghq.com/account_management/billing/custom_metrics/,
+///
+/// 100 seems like a reasonable place to start warning for now
+pub const DEFAULT_TAG_THRESHOLD: usize = 100;
+
+pub(crate) struct TrackerState {
+    /// Threshold at which to take the user defined action, and stop tracking
+    cardinality_threshold: usize,
+
+    /// The number of unique sets of tags seen for each metric
+    cardinality_count: usize,
+
+    /// For each metric, store cardinality
+    seen: BTreeMap<String, BTreeSet<BTreeSet<String>>>,
+
+    /// Actions to take when the threshold is exceeded
+    actions: Vec<ThresholdAction>,
+}
+
+impl TrackerState {
+    fn update(mut state: MutexGuard<'_, Self>, dd: &Datadog, metric: &str, tags: &[String]) {
+        debug_assert!(
+            !tags.is_empty(),
+            "update called with empty tags - should have been caught by Tracker::track"
+        );
+
+        let seen_tag_sets = match state.seen.get_mut(metric) {
+            Some(seen_tag_sets) => seen_tag_sets,
+            None => {
+                // This isn't very efficient, but because this code path is only taken once per metric, it's not a big deal.
+                // Doing this is far better than having to allocate the metric to a String for every call.
+                state.seen.insert(metric.to_string(), Default::default());
+                state.seen.get_mut(metric).unwrap()
+            }
+        };
+
+        // Is this set of tags new for this metric?
+        let set_is_novel = seen_tag_sets
+            .iter()
+            .all(|tag_set| tag_set.len() != tags.len() || tags.iter().any(|tag| !tag_set.contains(tag)));
+
+        if set_is_novel {
+            seen_tag_sets.insert(BTreeSet::from_iter(tags.iter().cloned()));
+            state.cardinality_count += 1;
+
+            // Check if we've exceeded the threshold
+            if state.cardinality_count >= state.cardinality_threshold {
+                Self::do_actions(state, dd, metric, tags);
+            }
+        }
+    }
+
+    fn do_actions(mut lock: MutexGuard<'_, Self>, dd: &Datadog, metric: &str, tags: &[String]) {
+        let actions = core::mem::take(&mut lock.actions);
+        if actions.is_empty() {
+            debug_assert!(
+                false,
+                "do_actions called with no actions - should have been caught by Tracker::state or Tracker::new"
+            );
+            return;
+        }
```

Comment:
I think we can get rid of the guard entirely imho... even if `actions` is empty, nothing harmful happens.

Reply:
Yes but up the stack we had to do a lot more work to get here (in `TrackerState::update`). The debug assertions ensures we don't accidentally write in control flow that causes unnecessary work to be done.

Reply:
If there are no actions the tracker should be constructed with no state - `Tracker::new` already ensures this is the case, and we **should** be dropping the internal state object entirely when we've fired the actions, however, I see we're no longer doing that since we've switched away from an enum-based state machine representation of the trackers internal state. I'm going to spend some time refactoring to ensure we drop things correctly and that these invalid states aren't representable, and then checks like this shouldn't be necessary.

Reply:
Removed